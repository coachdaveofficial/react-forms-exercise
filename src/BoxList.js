import React, { useState } from 'react';
import { v4 as uuid } from 'uuid';
import Box from './Box.js'
import NewBoxForm from './NewBoxForm.js'

function BoxList({initialBoxes = []}) {
    const [boxes, setBoxes] = useState(initialBoxes);

    const addBox = (box) => {
        let newBox = {...box, id: uuid() };
        setBoxes([...boxes, newBox]);
    }
    const removeBox = (id) => setBoxes(boxes.filter(b => b.id !== id));

    return (
        <div>
            <NewBoxForm addBox={addBox}/>
            {boxes.map(({ bgColor, width, height, id }) => {
                return (
                <div key={id}>
                    <Box bgColor={bgColor} width={width} height={height} remove={removeBox} id={id} />
                </div>
                )
            })}
        </div>
    );
}

export default BoxList;