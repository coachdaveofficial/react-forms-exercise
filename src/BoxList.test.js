import React from "react";
import { render, fireEvent } from "@testing-library/react";
import BoxList from "./BoxList";

it("renders without crashing", function () {
    render(<BoxList />)
});

it("matches snapshot", function () {
    const { asFragment } = render(<BoxList />);
    expect(asFragment()).toMatchSnapshot();
});

it('renders the list of boxes', () => {
    const boxes = [
        { id: '1', bgColor: 'red', width: 100, height: 200 },
        { id: '2', bgColor: 'blue', width: 150, height: 250 },
    ];
    const { getAllByTestId } = render(<BoxList initialBoxes={boxes} />);

    const boxElements = getAllByTestId('Box');

    expect(boxElements.length).toBe(boxes.length);
});

it('adds a new box when the form is submitted', () => {
    const initialBoxes = [
        { id: '1', bgColor: 'blue', width: 100, height: 200 },
    ];
    const {getByLabelText, getByText, getAllByTestId} = render(<BoxList initialBoxes={initialBoxes} />);

    const bgColorInput = getByLabelText('Background Color:');
    const widthInput = getByLabelText('Width:');
    const heightInput = getByLabelText('Height:');
    const addButton = getByText('Add Box');

    const newBox = {
        bgColor: 'red',
        width: '120',
        height: '180',
    };

    fireEvent.change(bgColorInput, { target: { value: newBox.bgColor } });
    fireEvent.change(widthInput, { target: { value: newBox.width } });
    fireEvent.change(heightInput, { target: { value: newBox.height } });
    fireEvent.click(addButton);

    const boxElements = getAllByTestId('Box');

    expect(boxElements.length).toBe(initialBoxes.length + 1);
    expect(boxElements[boxElements.length - 1]).toHaveStyle({
        backgroundColor: newBox.bgColor,
        width: `${newBox.width}px`,
        height: `${newBox.height}px`,
    });
});

it('removes a box when the "Remove me" button is clicked', () => {
    const boxes = [
        { id: '1', bgColor: 'red', width: 100, height: 200 },
        { id: '2', bgColor: 'blue', width: 150, height: 250 },
    ];
    const {getAllByTestId, getAllByText} = render(<BoxList initialBoxes={boxes} />);

    const removeButtons = getAllByText('Remove me');
    fireEvent.click(removeButtons[0]);

    const boxElements = getAllByTestId('Box');

    expect(boxElements.length).toBe(boxes.length - 1);
});