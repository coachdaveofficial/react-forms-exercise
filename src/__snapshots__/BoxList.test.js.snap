// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div>
    <form>
      <label
        for="bgColor"
      >
        Background Color: 
      </label>
      <input
        id="bgColor"
        name="bgColor"
        placeholder="background color"
        type="text"
        value=""
      />
      <label
        for="width"
      >
        Width: 
      </label>
      <input
        id="width"
        name="width"
        placeholder="width"
        type="text"
        value=""
      />
      <label
        for="height"
      >
        Height: 
      </label>
      <input
        id="height"
        name="height"
        placeholder="height"
        type="text"
        value=""
      />
      <button>
        Add Box
      </button>
    </form>
  </div>
</DocumentFragment>
`;
